### DISCUSSION TITLE
Compile error H4

### DISCUSSION THREAD

**Discussion by graemejw** [kachurovskiy/nanoels#125] [General]
Hi, just tried to compile the H4 code and I'm getting the following error

 error: suggest parentheses around '&&' within '||' [-Werror=parentheses]
   if (on && isPassMode() && (needZStops() && (z.leftStop == LONG_MAX || z.rightStop == LONG_MIN) || x.leftStop == LONG_MAX || x.rightStop == LONG_MIN)) {
                              ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


Compilation error: suggest parentheses around '&&' within '||' [-Werror=parentheses]

**Comment by kachurovskiy** [comment]
Hi @graemejw, what IDE are you using? Try Arduino 2 with target device set to `ESP32S3 Dev Module` as below:

![image](https://github.com/kachurovskiy/nanoels/assets/517919/0760df6b-0b00-4c72-bad2-92a348330ef5)

Full instructions on programming: https://github.com/kachurovskiy/nanoels/blob/main/h4/README.md#programming-the-controller

Cheers!

**Reply by graemejw** [reply]
Hi, I am using IDE 2.1.0 and have the ESP32S3 Dev Module selected[
<img width="1440" alt="Arduino-nanoels4" src="https://github.com/kachurovskiy/nanoels/assets/24543856/4aca47c6-e56c-40b0-8c21-1d1e39532995">
](url)

**Reply by kachurovskiy** [reply]
Looks like I need to update myself to 2.1.0 and fix those, hang on.

**Reply by kachurovskiy** [reply]
Hmm, I don't see any of those on 2.1.1:

![image](https://github.com/kachurovskiy/nanoels/assets/517919/6a0f788c-128a-42d1-909d-fea37c621f87)

Will try to find 2.1.0 binary to try to reproduce - or perhaps it only happens on Macs? Will keep looking.

**Reply by graemejw** [reply]
Hi thanks for the quick response, i've upgraded to 2.1.1 and still getting the same error

**Comment by kachurovskiy** [comment]
Looks like you have `Compiler warnings` set to `All` in Preferences - try setting it to `None`, it should help.

![image](https://github.com/kachurovskiy/nanoels/assets/517919/746429c1-2b2c-43b7-a06f-8fad3414fc0e)

Meanwhile I'll set mine to `All` and make those warnings go away :)

**Reply by graemejw** [reply]
Hi, I had jsut changed compiler warnings to default when you replied, with set at default, I no longer get the error, thanks for the help.
How close are you to producing the routines to machine full spheres?

**Reply by kachurovskiy** [reply]
You can do this today, although the functionality is still experimental:

1. Open a graphics editor, e.g. figma.com and draw your part with 1px = 0.01mm, below is an example of a sphere with diameter 10mm: ![image](https://github.com/kachurovskiy/nanoels/assets/517919/f69807ec-140a-45b1-8472-045d542e7975)
2. Download PNG of your frame using the `Export` menu on the right: 
![image](https://github.com/kachurovskiy/nanoels/assets/517919/132e6101-0add-43ea-b49a-1ed11867e692)
3. Open https://kachurovskiy.github.io/lathecode/ and import your image using `From image`
4. At the beginning of your lathecode, specify your tool e.g. `TOOL ROUND R4` for a 8mm round insert or `TOOL RECT R0.4 L3` for MGMN300 insert.
5. Scroll to the end of your lathecode, specify cutting off if you like by adding `L3` if your cutting tool width is 3
6. Zero the tool on the stock so that Z0 is when the tool touches stock from the right
7. Zero the tool on the stock so that X0 is on the centerline - take a light cut, measuring resulting stock diameter, entering half of it using the numpad and pressing `X`
8. Switch to `GCODE` mode

After generating GCode, make sure the toolpath makes sense. If your stock is bigger than the part, add `STOCK D31` at the top with your stock diameter. You can speed things up by adding lines `DEPTH CUT1` and `FEED MOVE500 PASS200 PART50` at the top:

![image](https://github.com/kachurovskiy/nanoels/assets/517919/8bf005ed-20da-4d2d-97eb-94581d80b2ca)

To run the GCode, turn on the spindle and click `Send to NanoEls H4` in the browser.

Should anything move unexpectedly, use the physical ![IconStop](https://github.com/kachurovskiy/nanoels/assets/517919/dddd0998-e25e-4d70-9b02-27a94a54e669) button on the H4 or the emergency stop.

Instead of just drawing a sphere, in Figma you can design any part you like and cut it in the same manner. During processing, image is split in half vertically and only the right half is taken for the tool path generation so if you're drawing a complex part you can just draw half of it :)

You can find the video of the cutting process and my first sphere at https://t.me/MaximKachurovskiy :) Cheers!

**Reply by graemejw** [reply]
Ok,that looks pretty cool, I am hoping to install the steppers on my lathe this weekend although I am still waiting on the drivers to arrive so it will be a while before I get everything up and running. I am hoping to use this to create a set of ball handles for my Quorn tool and cutter grinder. This whole process looks very promising, thanks for a great looking product, keep up the great work

**Comment by graemejw** [comment]
Hi, ok, ive had a look at figma but I cannot see how to set the 1px = 0.01mm ratio, can you tell me how I do this?

**Reply by kachurovskiy** [reply]
Using the Frame tool in the top-left add a frame onto the page:

![image](https://github.com/kachurovskiy/nanoels/assets/517919/b288a34a-e4e2-4e58-91e8-3755f6af7f3f)

Select your new frame and in the right panel set frame width to stock diameter, height to your stock length assuming 1px = 0.01mm - so for 10mm*30mm stock you'll have 1000px*3000px:

![image](https://github.com/kachurovskiy/nanoels/assets/517919/abfb20a0-f6d7-4f07-b987-4736ebfd3799)

Now you can use e.g. rectangle tool to specify your part in black within your frame:

![image](https://github.com/kachurovskiy/nanoels/assets/517919/fdbadc71-33f6-4620-a6c9-4768fee8a1e1)
